refactor to use tcprouterclient

xmonader opened the pull request:
- add support for multiple binaries (tcprouterserver, tcprouterclient)
- update project layout


still in the works:
- clean up logs
- register and retrieving services based on the secret

xmonader committed:
refactor to use tcprouterclient

xmonader committed:
if service has clientsecret check in the active connections

xmonader committed:
update gitignore and untrack wrongly commited binaries

zaibon commented on cmd/server/server.go:
```diff
@@ -83,12 +84,8 @@ func main() {
-		listeningHTTPPort: cfg.Server.HTTPPort,
-	}
-	s := NewServer(serverOpts, kv, cfg.Server.Services)
+	serverOpts := tcprouter.NewServerOptions(cfg.Server.Host, cfg.Server.Port, cfg.Server.HTTPPort, cfg.Server.ClientsPort)
```
For option I kinda like the fact that you can just create a literal here. The NewServerOptions doesn't do anything special

zaibon commented on server.go:
```diff
@@ -28,11 +29,23 @@ func (o ServerOptions) TLSAddr() string {
+func (o ServerOptions) ClientsAddr() string {
+	return fmt.Sprintf("%s:%d", o.listeningAddr, o.listeningForClientsPort)
+}
+func NewServerOptions(listeningAddr string, listeningTLSPort, listeningHTTPPort, listeningForClientsPort uint) ServerOptions {
```
I would get rid of this one. All field are public anyhow so you can just create the object using a literal.

zaibon commented on server.go:
```diff
@@ -150,6 +165,48 @@ func (s *Server) serveTLS(ctx context.Context) {
+				continue
+			}
+			// TODO: need to track these connection and close then when needed
+			go s.handleTCPRouterClientConnection(conn)
```
No need to create goroutine here, you are just doing the handshake and saving the connection for later. This can happen in this block

zaibon commented on server.go:
```diff
@@ -150,6 +165,48 @@ func (s *Server) serveTLS(ctx context.Context) {
+	if err := hs.Read(br); err != nil {
+		log.Printf("handshake failed")
+		return err
+	} else {
```
no need for the else block since you return at the line before.

zaibon commented on server.go:
```diff
@@ -150,6 +165,48 @@ func (s *Server) serveTLS(ctx context.Context) {
+		log.Printf("handshake failed")
+		return err
+	} else {
+		log.Printf("handshake done %v", hs)
```
Before you validate the handshake you need to verify the magic number.

zaibon commented on server.go:
```diff
@@ -213,8 +270,29 @@ func (s *Server) handleService(mainconn net.Conn, serverName, peeked string, isT
+
+}
+
+func (s *Server) forwardToActiveConnection(conn, activeConn net.Conn) error {
```
`forwardToActiveConnection` and `forward` are nearly the same function, can't we just move the dial out of `forward` and delete `forwardToActiveConnection` ?

zaibon commented on Makefile:
```diff
@@ -0,0 +1,10 @@
+server:
```
usually when I have multiple binaries to build in a repo, I like that the output of the build endup in a `bin` directory. 
So it doesn't polute the root and every binaries are in the same spot

zaibon reviewed (changes requested):
Logic looks good. Just some style comments

xmonader committed:
Move resulting binaries into bin dir

xmonader committed:
move server options to literal instead of using NewServerOptions

xmonader committed:
refactor forwarding connection

zaibon commented on handshake_test.go:
```diff
@@ -68,3 +68,20 @@ func TestHandshake(t *testing.T) {
 	wg.Wait()
 }
+
+func TestHandshakeAgainstTCPRouterServer(t *testing.T) {
```
This test will always fail, what's the point of having it ?

zaibon replied:
This test will always fail, what's the point of having it ?

zaibon commented on server.go:
```diff
@@ -150,6 +158,50 @@ func (s *Server) serveTLS(ctx context.Context) {
+		log.Printf("handshake failed")
+		return err
+	}
+	if hs.MagicNr == MagicNr {
```
usually we do the check the other way, so we save one level of indentation
```
if hs.MagicNR != MagicNR {
    //early return
   return
}

s.activeConnections[string(hs.Secret[:])] = mainconn
``

zaibon commented on server.go:
```diff
@@ -213,31 +265,44 @@ func (s *Server) handleService(mainconn net.Conn, serverName, peeked string, isT
-		log.Printf("failed to forward traffic: %v\n", err)
+	if service.ClientSecret != "" {
+		// forward to an active connection
+		if err := s.forwardConnection(conn, s.activeConnections[service.ClientSecret]); err != nil {
```
same remark as before, delay the error handling after the if/else block so you only do it once.
Here you don't print the error for the line 271 for example.

xmonader committed:
remove obsolete test

xmonader committed:
update PR Change requests

zaibon committed:
more improvement and cleanup

- stop forwarding when one of the connection closes
- make the client loop forever
- improve handshake serialization

zaibon committed:
fix parsing of HTTP header

zaibon committed:
use the proper config to connect to store backend

xmonader committed:
Merge pull request #11 from zaibon/master-clientsupport

more improvement and cleanup
